04-onion-routing.md

```diff
@@ -661,121 +662,141 @@ The top byte of `failure_code` can be read as a set of flags:
-  report an error:
-    1. type: 21 (`expiry_too_far`)
-  - if the channel is disabled, the current channel setting for the outgoing
-  channel are reported:
-    1. type: UPDATE|20 (`channel_disabled`)
-    2. data:
-       * [`2`: `flags`]
-       * [`2`:`len`]
-       * [`len`:`channel_update`]
+
+If the onion `version` byte is unknown:
+1. type: BADONION|PERM|4 (`invalid_onion_version`)
+2. data:
+   * [`32`:`sha256_of_onion`]
+
+If the onion HMAC is incorrect:
+1. type: BADONION|PERM|5 (`invalid_onion_hmac`)
+2. data:
+   * [`32`:`sha256_of_onion`]
+
+If the ephemeral key in the onion is unparsable:
+1. type: BADONION|PERM|6 (`invalid_onion_key`)
+2. data:
+   * [`32`:`sha256_of_onion`]
+
+If an otherwise unspecified, transient error occurs during forwarding to its
+receiving peer (e.g. channel capacity reached, too many in-flight HTLCs, etc.):
+1. type: UPDATE|7 (`temporary_channel_failure`)
+2. data:
+   * [`2`:`len`]
+   * [`len`:`channel_update`]
+
+If an otherwise unspecified, permanent error occurs during forwarding to its
+receiving peer (e.g. channel recently closed):
+1. type: PERM|8 (`permanent_channel_failure`)
+
+If the outgoing channel has requirements advertised in its
+`channel_announcement` `features` which were NOT included in the onion:
+1. type: PERM|9 (`required_channel_feature_missing`)
+
+If the receiving peer specified by the onion is NOT known:
+1. type: PERM|10 (`unknown_next_peer`)
+
+If the HTLC amount is less than the currently specified minimum amount, the
+amount of the incoming HTLC and the current channel setting for the outgoing
+channel [FIXME: is there a more concise way to say this?] are reported:
+1. type: UPDATE|11 (`amount_below_minimum`)
+2. data:
+   * [`8`:`htlc_msat`]
+   * [`2`:`len`]
+   * [`len`:`channel_update`]
+
+If the HTLC does NOT pay a sufficient fee, the amount of the incoming HTLC and
+the current channel setting for the outgoing channel are reported:
+1. type: UPDATE|12 (`fee_insufficient`)
+2. data:
+   * [`8`:`htlc_msat`]
+   * [`2`:`len`]
+   * [`len`:`channel_update`]
+
+If the `outgoing_cltv_value` does NOT match the `update_add_htlc`'s
+`cltv_expiry` minus the `cltv_expiry_delta` for the outgoing channel, the
+`cltv_expiry` and the current channel setting for the outgoing channel are
+reported:
+1. type: UPDATE|13 (`incorrect_cltv_expiry`)
+2. data:
+   * [`4`:`cltv_expiry`]
+   * [`2`:`len`]
+   * [`len`:`channel_update`]
+
+If the `cltv_expiry` is too near, the current channel setting for the
+outgoing channel are reported:
+1. type: UPDATE|14 (`expiry_too_soon`)
+2. data:
+   * [`2`:`len`]
+   * [`len`:`channel_update`]
+
+If the `cltv_expiry` is unreasonably far in the future, its also possible to
+report an error:
+1. type: 21 (`expiry_too_far`)
+
+If the channel is disabled, the current channel setting for the outgoing
+channel are reported:
+1. type: UPDATE|20 (`channel_disabled`)
+2. data:
+   * [`2`: `flags`]
+   * [`2`:`len`]
+   * [`len`:`channel_update`]
```

Comment:
For future work, I'd restore the text you had, just move the definitions of the structures into its own section, then turn the words describing each one into requirements.  But let's leave for now!

Reply:
Ah I get your meaning now, I will tackle this in next PR

Reply:
Left myself a couple [TODO] tags